### PR TITLE
Void endpoint

### LINKED ISSUE #11: Implement the Deferred Void endpoint
https://developers.afterpay.com/afterpay-online/reference#void-payment

### PR DESCRIPTION
**Pull request by garciajordy**
# Void Endpoint

Add a class for the Void endpoint taking the `order_id` to add it to the URL of the endpoint.

Add the attr_accessor `amount` to the payment component as it is needed for the void class.

Add tests for the void class when giving the correct order_id and the wrong order_id.

fix #11

### CODE REVIEW

File: lib/afterpay.rb [hunk 1]
```diff
@@ -8,6 +8,7 @@
 require 'afterpay/api/payment'
 require 'afterpay/api/payment/capture'
 require 'afterpay/api/payment/refund'
+require 'afterpay/api/payment/void'
```
Comment (ChristianRimondi) [inline lib/afterpay.rb:11]: you should move this `require` in the previous commit

File: lib/afterpay/components/payment.rb [hunk 2]
```diff
@@ -12,6 +12,11 @@ class Payment < Base
       # @return [String]
       # The merchant's order id / reference that this payment corresponds to
       attr_accessor :merchant_reference
+
+      # @attribute amount
+      # @return [Afterpay::Components::Money]
+      # The refund amount.
+      attr_accessor :amount
```
Comment (ChristianRimondi) [inline lib/afterpay/components/payment.rb:19]: Same here, or create a different commit

File: spec/api/payment/void_spec.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe Afterpay::API::Payment::Void do
+  describe '#call' do
+    let(:token) { 'some_token' }
```
Comment (ChristianRimondi) [inline spec/api/payment/void_spec.rb:7]: You don't need the token in the void request, only the `amount` https://developers.afterpay.com/afterpay-online/reference#void-payment

You can use [this spec](https://github.com/nebulab/afterpay-ruby/blob/main/spec/api/payment/refund_spec.rb) as a reference

File: spec/api/payment/void_spec.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        stub_request(:post, %r{api.us.afterpay.com/v2/payments/#{order_id}/void})
+          .to_return(
+            status: 201,
+            body: raw_response,
+            headers: { 'Content-Type' => 'application/json' }
+          )
+      end
+
+      it 'returns hashie mash' do
+        expect(described_class.call(payment: payment, order_id: order_id, amount: "12.11").body).to be_instance_of Hashie::Mash
```
Comment (ChristianRimondi) [inline spec/api/payment/void_spec.rb:24]: the `amount` should be a `Afterpay::Components::Money` and should be included in the payment object.

File: spec/api/payment/void_spec.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      end
+
+      it 'returns afterpay order id' do
+        expect(described_class.call(payment: payment, order_id: order_id, amount: "12.11").body).to include(:paymentState)
```
Comment (ChristianRimondi) [inline spec/api/payment/void_spec.rb:28]: the amount should be a `Afterpay::Components::Money` and should be included in the payment object.

File: spec/api/payment/void_spec.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        stub_request(:post, %r{api.us.afterpay.com/v2/payments/12/void})
+          .to_return(
+            status: 404,
+            body: raw_response,
+            headers: { 'Content-Type' => 'application/json' }
+          )
+      end
+
+      it 'raises NotFoundError' do
+        expect { described_class.call(payment: payment, order_id: "12", amount: "12.11") }.to raise_error Afterpay::NotFoundError
```
Comment (ChristianRimondi) [inline spec/api/payment/void_spec.rb:53]: the amount should be a `Afterpay::Components::Money` and should be included in the payment object.

File: spec/api/payment/void_spec.rb [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+      end
+
+      it 'returns proper error message' do
+        expect { described_class.call(payment: payment, order_id: "12", amount: "12.11") }.to raise_error error_message
```
Comment (ChristianRimondi) [inline spec/api/payment/void_spec.rb:57]: the amount should be a `Afterpay::Components::Money` and should be included in the payment object.

File: spec/api/payment/void_spec.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        {
+          errorCode: 'invalid',
+          errorId: '5393ff621317c438',
+          message: error_message,
+          httpStatusCode: 404
+        }.to_json
+      end
+
+      before(:each) do
+        stub_request(:post, %r{api.us.afterpay.com/v2/payments/12/void})
```
Comment (ChristianRimondi) [inline spec/api/payment/void_spec.rb:44]: Can you create an `order_id` variable also in this context? Like you do in the `with proper params` context

**Review by ChristianRimondi — CHANGES_REQUESTED** [review]

File: lib/afterpay.rb (hunk 1 above)
Comment (garciajordy) [reply lib/afterpay.rb:11]: Done!

File: lib/afterpay/components/payment.rb (hunk 2 above)
Comment (garciajordy) [reply lib/afterpay/components/payment.rb:19]: done!

File: spec/api/payment/void_spec.rb (hunk 3 above)
Comment (garciajordy) [reply spec/api/payment/void_spec.rb]: Removed the token as it is only needed in the Auth

File: spec/api/payment/void_spec.rb (hunk 4 above)
Comment (garciajordy) [reply spec/api/payment/void_spec.rb]: Added the amount as a money object like requested!

File: spec/api/payment/void_spec.rb (hunk 5 above)
Comment (garciajordy) [reply spec/api/payment/void_spec.rb]: Added the amount as a money object like requested!

File: spec/api/payment/void_spec.rb (hunk 6 above)
Comment (garciajordy) [reply spec/api/payment/void_spec.rb]: Added the amount as a money object like requested!

File: spec/api/payment/void_spec.rb (hunk 7 above)
Comment (garciajordy) [reply spec/api/payment/void_spec.rb]: Added the amount as a money object like requested!

File: spec/api/payment/void_spec.rb (hunk 8 above)
Comment (garciajordy) [reply spec/api/payment/void_spec.rb]: Done created an order_id variable with a different number

**Review by ChristianRimondi — APPROVED** [review]
LGTM 👍

**Pull request merged**